optimum/habana/transformers/models/llama/modeling_llama.py

```diff
@@ -798,6 +820,14 @@ def prepare_inputs_for_generation(
         else:
             model_inputs = {"input_ids": input_ids}
 
+        bucket_size = kwargs.get("bucket_size")
+        cache_len = self.get_cache_len()
+        cache_idx = None
+        if cache_len and bucket_size > 0:
+            idx = torch.div(token_idx - 1, bucket_size, rounding_mode="floor")
+            if idx < (cache_len // bucket_size):
+                cache_idx = (idx.item() + 1) * bucket_size
```

Comment:
@x574chen  - Also one more query here that can we move lines 823 to 829 to utils.py only and pass cache_idx in kwargs and just have one line here that cache_idx = kwargs.get("cache_idx"). It will make the bucketing changes easier for other models. Sorry for all these late review comments.

Reply:
Rest all changes look good to me ; basically what I m thinking is that we just pass cache_idx in modeling_llama.py and just have the additional change of slicing the KV cache in the attention block code.

Reply:
The .item() is used to ensure that cache_idx (one of model inputs) is an integer, not a tensor. This prevents HPUGraph from calling replay when the value of cache_idx changes. Please correct me if my understanding of the use of HPU graph is incorrect.